libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs

```diff
@@ -510,50 +529,74 @@ unUnifyToVMap uview = case uview of
+compactRewardMap :: UMap c -> Map (Credential 'Staking c) (CompactForm Coin)
 compactRewardMap x = Map.map rdReward $ unUnify $ RewDepUView x
 
 -- | Extract a deposits `Map` from a 'UMap'
-depositMap :: UMap c -> Map.Map (Credential 'Staking c) Coin
+depositMap :: UMap c -> Map (Credential 'Staking c) Coin
 depositMap x = Map.map (fromCompact . rdDeposit) $ unUnify $ RewDepUView x
 
 -- | Extract a pointers `Map` from a 'UMap'
-ptrMap :: UMap c -> Map.Map Ptr (Credential 'Staking c)
+ptrMap :: UMap c -> Map Ptr (Credential 'Staking c)
 ptrMap x = unUnify $ PtrUView x
 
 -- | Extract a pointers `Map` from a 'UMap'
-revPtrMap :: UMap c -> Map.Map (Credential 'Staking c) (Set Ptr)
+revPtrMap :: UMap c -> Map (Credential 'Staking c) (Set Ptr)
 revPtrMap UMap {umElems} =
   Map.foldlWithKey'
     (\ans k (UMElem _ ptrSet _ _) -> if Set.null ptrSet then ans else Map.insert k ptrSet ans)
     Map.empty
     umElems
 
 -- | Extract a stake pool delegations `Map` from a 'UMap'
-sPoolMap :: UMap c -> Map.Map (Credential 'Staking c) (KeyHash 'StakePool c)
+sPoolMap :: UMap c -> Map (Credential 'Staking c) (KeyHash 'StakePool c)
 sPoolMap x = unUnify $ SPoolUView x
 
 -- | Extract a delegated-representatives `Map` from a 'UMap'
-dRepMap :: UMap c -> Map.Map (Credential 'Staking c) (Credential 'Voting c)
+dRepMap :: UMap c -> Map (Credential 'Staking c) (Credential 'Voting c)
 dRepMap x = unUnify $ DRepUView x
 
 -- | Extract a domain-restricted `Map` of a `UMap`.
 -- If `Set k` is small this should be efficient.
-domRestrictedMap :: Set k -> UView c k v -> Map.Map k v
+domRestrictedMap :: Set k -> UView c k v -> Map k v
 domRestrictedMap setk = \case
   RewDepUView UMap {umElems} -> Map.mapMaybe umElemRDPair (Map.restrictKeys umElems setk)
   PtrUView UMap {umPtrs} -> Map.restrictKeys umPtrs setk
   SPoolUView UMap {umElems} -> Map.mapMaybe umElemSPool (Map.restrictKeys umElems setk)
   DRepUView UMap {umElems} -> Map.mapMaybe umElemDRep (Map.restrictKeys umElems setk)
 
+toStakeCredentials :: UMap c -> StakeCredentials c
+toStakeCredentials umap =
+  StakeCredentials
+    { scRewards = rewardMap umap
+    , scDeposits = depositMap umap
+    , scSPools = sPoolMap umap
+    , scDReps = dRepMap umap
+    , scPtrs = ptrMap umap
+    , scPtrsInverse = revPtrMap umap
+    }
+
+domRestrictedStakeCredentials :: Set (Credential 'Staking c) -> UMap c -> StakeCredentials c
+domRestrictedStakeCredentials setk UMap {umElems, umPtrs} =
+  let umElems' = Map.restrictKeys umElems setk
+      ptrs = Map.mapMaybe umElemPtrs umElems'
+   in StakeCredentials
+        { scRewards = Map.mapMaybe (\e -> fromCompact . rdReward <$> umElemRDPair e) umElems'
+        , scDeposits = Map.mapMaybe (\e -> fromCompact . rdDeposit <$> umElemRDPair e) umElems'
+        , scSPools = Map.mapMaybe umElemSPool umElems'
+        , scDReps = Map.mapMaybe umElemDRep umElems'
+        , scPtrs = umPtrs `Map.restrictKeys` fold ptrs
+        , scPtrsInverse = ptrs
```

Comment:
This got me thinking that there is a `revPtrMap` somewhere in `UMap`, that should perhaps be called `invPtrMap` to keep it consistent between using `reverse` or `inverse` as the word to describe the map in the opposite direction? Just a thought.

Reply:
I was also wondering whether `UMap.domRestrict` could be used for constructing these. :thinking:

Reply:
Good idea about renaming to `invPtrMap`. Reverse implies ordering, which isn't applicable here.

I don't think `domRestrict` is applicable to this problem.

The deal about this domain restriction in `domRestrictedStakeCredentials` is to do it only once, because it is an expensive computation, which `domRestrict` would not allow us to avoid. Also `domRestrict` requires a Map and here we have a Set.